Introduce multistore technical documentation for configuration forms and settings (#843)
PrestaShop/docs: src/content/1.7/development/Multistore/shop-context/_index.md

---
title: Shops context
menuTitle: Shops context
summary: "How to get the current shop context"
weight: 20
---

# Shop context

> In many cases, your code will need to know if the multistore mode is enabled, if there are more than one shops activated, and if yes, what is the current shop context. This is where the shop context and multistore feature adapters will come in handy.

Reviewer: ```suggestion
In many cases, your code will need to know if the multistore mode is enabled, if there are more than one shop activated, and if yes, what is the current shop context. This is where the shop context and multistore feature adapters will come in handy.
```

## The multistore feature adapter

**Service identifier:** `prestashop.adapter.feature.multistore` (it takes the legacy configuration adapter as a parameter) <br>
**Class:** `PrestaShop\PrestaShop\Adapter\Feature\MultistoreFeature`

The multistore feature adapter can be used to:

- detect if multistore was enabled in the configuration,

> - detect if multistore is active (if it is enabled, and if there are more than one active shops),

Reviewer: ```suggestion
- detect if multistore is active (if it is enabled, and if there are more than one active shop),
```

- enable/disable the multistore feature.

Example:

```php
<?php

// Get the multistore feature adapter from the service container
$multistoreFeature = $this->get('prestashop.adapter.feature.multistore');

// Is the multistore feature enabled?
$isMultistoreEnabled = $multistoreFeature->isActive();

// This will check that multistore is enabled AND that there are at least two active shops
$isMultistoreUsed = $multistoreFeature->isUsed();

// Disable / enable multistore feature
$multistoreFeature->disable();
$multistoreFeature->enable();

```

## The shop context adapter

**Service identifier:** `prestashop.adapter.shop.context` <br>
**Class:** `PrestaShop\PrestaShop\Adapter\Shop\Context`

> When the multistore feature is enabled, at the top of all the back office pages, a dropdown appears, it allows selecting the shop context we want (all shops, a group shop, a single shop...). The shop context adapter will tell you in what context your code is operating.

Reviewer: ```suggestion
When the multistore feature is enabled, at the top of all back office pages, a dropdown appears, it allows selecting the shop context we want (all shops, a group shop, a single shop). The shop context adapter will tell you in what context your code is operating.
```

Here are some examples of commonly used functions in the shop context adapter:

```php
<?php

// Get the shop context adapter from the service container

> $shopContextAdapter = $this->get('prestashop.adapter.shop.context');

Reviewer: ```suggestion
$shopContext = $this->get('prestashop.adapter.shop.context');
```

shopContext or shopContextAdapter? :)

// Get a list of all active shops
$shopList = $shopContext->getShops(false, true);

> // Is the current context a group of shops ?

Reviewer: ```suggestion
// Is the current context a group of shops?
```

$isShopGroupContext = $shopContext->isShopGroupContext();

> // Is the current context a single shop ?

Reviewer: ```suggestion
// Is the current context a single shop?
```

$isSingleShopContext = $shopContext->isSingleShopContext();

// Is "all shops" context
$isAllShopContext = $shopContext->isAllContext();

> 

Reviewer: ```suggestion
```

// Get a list of IDs for all shops belonging to the current context (useful is the current context is a group) 
$shopList = $shopContext->getContextListShopID();

// Get the id of the shop in the current context (useful if the current context is a single shop)
$currentIdShop = $shopContext->getContextShopID();
```

> The shop context adapter class has many functions with self-explanatory naming and helpful comments, feel free to [explore it by yourself](https://github.com/PrestaShop/PrestaShop/blob/1.7.7.x/src/Adapter/Shop/Context.php).

Reviewer: ```suggestion
The shop context adapter class has many functions with self-explanatory naming and helpful comments, feel free to [explore it by yourself](https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/Shop/Context.php).
```

Not sure if we should link to specific version.

Author: Good question, @matks what do you think ? I thought since we are in 1.7.7 documentation it was better to link the 1.7.7 code, because the develop branch is ever changing.

Reviewer: It's global documentation, right? And this feature you're documenting in some places will work from 1.7.8, I think `develop` is better choice, also, if we change something on develop it should be documented anyway

Author: true

---
title: Shops context
menuTitle: Shops context
summary: "How to get the current shop context"
weight: 20
---

# Shop context

In many cases, your code will need to know if the multistore mode is enabled, if there are more than one shops activated, and if yes, what is the current shop context. This is where the shop context and multistore feature adapters will come in handy.

## The multistore feature adapter

> **Service identifier:** `prestashop.adapter.feature.multistore`  (it takes the legacy configuration adapter as a parameter) <br>

Reviewer 2: ```suggestion
**Service identifier:** `prestashop.adapter.feature.multistore` (it takes the legacy configuration adapter as a parameter) <br>
```

**Class:** `PrestaShop\PrestaShop\Adapter\Feature\MultistoreFeature`

The multistore feature adapter can be used to:

> - detect if multistore was enabled in the configuration
> 
> - detect if multistore is active (if it is enabled, and if there are more than one active shops)
> 
> - enable/disable the multistore feature

Reviewer 2: ```suggestion
- detect if multistore was enabled in the configuration,

- detect if multistore is active (if it is enabled, and if there are more than one active shops),

- enable/disable the multistore feature.
```

Example:

```php
<?php

> // get the multistore feature adapter from the service container

Reviewer 2: ```suggestion
// Get the multistore feature adapter from the service container
```

$multistoreFeature = $this->get('prestashop.adapter.feature.multistore');

> // Is the multistore feature enabled ?

Reviewer 2: ```suggestion
// Is the multistore feature enabled?
```

$isMultistoreEnabled = $multistoreFeature->isActive();

// This will check that multistore is enabled AND that there are at least two active shops
$isMultistoreUsed = $multistoreFeature->isUsed();

// Disable / enable multistore feature
$multistoreFeature->disable();
$multistoreFeature->enable();

```

## The shop context adapter

**Service identifier:** `prestashop.adapter.shop.context` <br>
**Class:** `PrestaShop\PrestaShop\Adapter\Shop\Context`

> When the multistore feature is enabled, at the top of all the back office pages, a dropdown appears, it allows selecting the shop context we want (all shops, a group shop, a single shop...). The shop context adapter will tell you in what context your are operating.

Reviewer 2: ```suggestion
When the multistore feature is enabled, at the top of all the back office pages, a dropdown appears, it allows selecting the shop context we want (all shops, a group shop, a single shop...). The shop context adapter will tell you in what context your shop is operating.
```

Author: I understand that you imply that when we are in multistore, each store belongs to a shop?  
I find it a bit ambiguous, what we want to know here is in which shop context the code is operating. 

I could replace "you are operating" by "your code is operating" though

Here are some examples of commonly used functions in the shop context adapter:

```php
<?php

> // get the shop context adapter from the service container

Reviewer 2: ```suggestion
// Get the shop context adapter from the service container
```

$shopContextAdapter = $this->get('prestashop.adapter.shop.context');

// get a list of all active shops
$shopList = $shopContext->getShops(false, true);

> // is the current context a group of shops ?

Reviewer 2: ```suggestion
// Is the current context a group of shops ?
```

$isShopGroupContext = $shopContext->isShopGroupContext();

> // is the current context a single shop ?

Reviewer 2: ```suggestion
// Is the current context a single shop ?
```

$isSingleShopContext = $shopContext->isSingleShopContext();

> // is "all shops" context

Reviewer 2: ```suggestion
// Is "all shops" context
```

$isAllShopContext = $shopContext->isAllContext();

> 
> // get a list of IDs for all shops belonging to the current context (useful is the current context is a group) 

Reviewer 2: ```suggestion
// Get a list of IDs for all shops belonging to the current context (useful is the current context is a group) 
```

$shopList = $shopContext->getContextListShopID();

> // get the id of the shop in the current context (useful if the current context is a single shop)

Reviewer 2: ```suggestion
// Get the id of the shop in the current context (useful if the current context is a single shop)
```

$currentIdShop = $shopContext->getContextShopID();

> 
> 

Reviewer 2: ```suggestion
```
